### PR TITLE
Introducing Module-NMC controller

### PR DESCRIPTION
**Pull request by yevgeny-shnaidman**
This PR introduces a new controller, that is responsible for reconciliation of NMC based on the data in Module and Node The following are the main points:
1) It watches for changes in Modules, Nodes (labels, becoming
   schedulable) and Jobs belonging owned by Modules (in order to
   reschedule once Build/Sign job are finished)
2) Per reconciled Module, go over all the nodes in the cluster
   and reconcile their NMCs based on the Module/Node state

Changes in the PR:
1) adding functions to filter package that are resposible for filtering
    the events (nodes. modules, jobs) and finding out which modules
    should be reconciled
2)  changing the NMC handler logic a little, adding verification for
    container image presence in case module should be enable.
3)  Module-NMC Reconciliation is being done both in case Kernel Mapping
    exists and in case it does not exists ( solves the garbage
    collection flow)

This PR does not include Module status updates based on NMC updates

### CODE REVIEW

File: cmd/manager/main.go [hunk 1]
```diff
@@ -146,11 +149,20 @@ func main() {
...
 	)
-
 	if err = mc.SetupWithManager(mgr, constants.KernelLabel); err != nil {
 		cmd.FatalError(setupLogger, err, "unable to create controller", "name", controllers.ModuleReconcilerName)
 	}
 
+	mnc := controllers.NewModuleNMCReconciler(
+		client,
+		kernelAPI,
+		nmcHandler,
```
Comment (ybettan) [inline cmd/manager/main.go:159]: Should this one be renamed to `nmcAPI` to be consistent with the rest of the API passed to it (or to the Module reconciler)?

File: controllers/module_nmc_reconciler.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+//+kubebuilder:rbac:groups="core",resources=nodes,verbs=get;watch;patch
```
Comment (ybettan) [inline controllers/module_nmc_reconciler.go:25]: Why do we need to `patch` nodes?

File: controllers/module_nmc_reconciler.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+//+kubebuilder:rbac:groups=kmm.sigs.x-k8s.io,resources=nodemodulesconfigs,verbs=get;list;watch;update;patch;create
```
Comment (ybettan) [inline controllers/module_nmc_reconciler.go:26]: What is the difference between `update` and `patch`?

File: controllers/module_nmc_reconciler.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+		reconHelper: reconHelper,
+		nmcHandler:  nmcHandler,
+	}
+}
+
+func (mnr *ModuleNMCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// get the reconciled module
+	logger := log.FromContext(ctx)
+
+	logger.Info("Starting Module-NMS reconcilation", "module name and namespace", req.NamespacedName)
```
Comment (ybettan) [inline controllers/module_nmc_reconciler.go:56]: Who says that the reconciled object is a `Module`? Can't it be a `Node`? a `Job`?

File: controllers/module_nmc_reconciler.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+	mod, err := mnr.reconHelper.getRequestedModule(ctx, req.NamespacedName)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			logger.Info("Module deleted")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get the requested %s KMMO CR: %w", req.NamespacedName, err)
```
Comment (ybettan) [inline controllers/module_nmc_reconciler.go:64]:
```suggestion
		return ctrl.Result{}, fmt.Errorf("failed to get the requested %s KMMO CR: %v", req.NamespacedName, err)
```

Are we using the wrapped error anywhere?

File: controllers/module_nmc_reconciler.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+	// get all nodes
+	nodes, err := mnr.reconHelper.getNodesList(ctx)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get nodes list: %v", err)
+	}
+
+	errs := make([]error, 0, len(nodes))
+	for _, node := range nodes {
+		kernelVersion := strings.TrimSuffix(node.Status.NodeInfo.KernelVersion, "+")
+		mld, err := mnr.kernelAPI.GetModuleLoaderDataForKernel(mod, kernelVersion)
```
Comment (ybettan) [inline controllers/module_nmc_reconciler.go:76]: Why do we need to do that for every node? Can we extract it outside of the loop?

File: controllers/module_nmc_reconciler.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+		if err != nil && !errors.Is(err, module.ErrNoMatchingKernelMapping) {
+			logger.Info(utils.WarnString(fmt.Sprintf("internal errors while fetching kernel mapping for version %s: %v", kernelVersion, err)))
+			errs = append(errs, err)
+			continue
+		}
+
+		err = mnr.nmcHandler.HandleModuleNMC(ctx, mod, mld, node)
+		errs = append(errs, err)
```
Comment (ybettan) [inline controllers/module_nmc_reconciler.go:84]: Aren't you appending `nil` errors here as well?

File: cmd/manager/main.go (hunk 1 above)
Comment (yevgeny-shnaidman) [reply cmd/manager/main.go:159]: This is a controller, we are not creating/returning an Interface ( API), but the struct itself, as the rest of the reconcilers/controllers, that's why there is no API in the name

File: controllers/module_nmc_reconciler.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+		return ctrl.Result{}, fmt.Errorf("failed to reconcile module %s/%s with nodes: %v", mod.Namespace, mod.Name, err)
+	}
+	return ctrl.Result{}, nil
+}
+
+//go:generate mockgen -source=module_nmc_reconciler.go -package=controllers -destination=mock_module_nmc_reconciler.go moduleNMCReconcilerHelperAPI
+
+type moduleNMCReconcilerHelperAPI interface {
+	getRequestedModule(ctx context.Context, namespacedName types.NamespacedName) (*kmmv1beta1.Module, error)
+	getNodesList(ctx context.Context) ([]v1.Node, error)
```
Comment (ybettan) [inline controllers/module_nmc_reconciler.go:98]: I am wondering why `getRequestedModule` isn't a method in the `module` package and why we don't have a `node` package that will hold `getNodeList` as well.

File: controllers/module_nmc_reconciler.go (hunk 3 above)
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go:26]: that depends on the action that you execute on the object. I prefer to mention both of them, so that we can execute them in the code without a need to update the rules

File: controllers/module_nmc_reconciler.go [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+	client client.Client
+}
+
+func newModuleNMCReconcilerHelper(client client.Client) moduleNMCReconcilerHelperAPI {
+	return &moduleNMCReconcilerHelper{
+		client: client,
+	}
+}
+
+func (mnrh *moduleNMCReconcilerHelper) getRequestedModule(ctx context.Context, namespacedName types.NamespacedName) (*kmmv1beta1.Module, error) {
```
Comment (ybettan) [inline controllers/module_nmc_reconciler.go:111]: Isn't this function a dup of `getRequestedModule` in the `module_reconciler.go` file?

File: controllers/module_nmc_reconciler.go [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+	mod := kmmv1beta1.Module{}
+
+	if err := mnrh.client.Get(ctx, namespacedName, &mod); err != nil {
+		return nil, fmt.Errorf("failed to get the kmmo module %s: %w", namespacedName, err)
+	}
+	return &mod, nil
+}
+
+func (mnrh *moduleNMCReconcilerHelper) getNodesList(ctx context.Context) ([]v1.Node, error) {
```
Comment (ybettan) [inline controllers/module_nmc_reconciler.go:120]: Same here, this function exist (almost) in `module_reconciler.go`

File: cmd/manager/main.go (hunk 1 above)
Comment (ybettan) [reply cmd/manager/main.go:159]: `nmcHandler` is a public interface. Why don't we call it `nmcAPI` instead?

File: controllers/module_nmc_reconciler.go (hunk 4 above)
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go:56]: because this is how this reconciler is defined. It is defined For Module, and for every "watch" it finds the appropriate Module to reconcile

File: controllers/module_nmc_reconciler.go (hunk 5 above)
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go:64]: Fixed. We do use wrapped errors , when needed. For example in the KernelMapper, to get the appropriate MLD, if exists

File: controllers/module_nmc_reconciler.go (hunk 6 above)
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go:76]: because the kernel version of the nodes may differ

File: controllers/module_nmc_reconciler.go (hunk 7 above)
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go:84]: yes, but errors.Join knows how to handle them and ignore

File: controllers/module_nmc_reconciler.go (hunk 8 above)
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go:98]: As discussed in Slack, i am leaving it here and we can decide on it and the rest of functions in packages in v2.1

File: controllers/module_nmc_reconciler.go (hunk 9 above)
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go:111]: As discussed in Slack, i am leaving it here and we can decide on it and the rest of functions in packages in v2.1

File: controllers/module_nmc_reconciler.go (hunk 10 above)
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go:120]: As discussed in Slack, i am leaving it here and we can decide on it and the rest of functions in packages in v2.1

File: cmd/manager/main.go (hunk 1 above)
Comment (yevgeny-shnaidman) [reply cmd/manager/main.go]: moved the NMC-Handler API into the controller, since it implements controller logic

File: controllers/module_nmc_reconciler.go (hunk 5 above)
Comment (ybettan) [reply controllers/module_nmc_reconciler.go]: I meant, "do we use the wrapped error in this case". If you fixed it I assume we were not using it :)

File: controllers/module_nmc_reconciler.go (hunk 6 above)
Comment (ybettan) [reply controllers/module_nmc_reconciler.go:79]: But I don't see `node` being passed to this function or the implementation depending on the nodes in any way. Am I missing something?

File: controllers/module_nmc_reconciler.go (hunk 7 above)
Comment (ybettan) [reply controllers/module_nmc_reconciler.go:96]: Won't it be clearer to only append real errors instead of delegating it to `Join`?

File: controllers/module_nmc_reconciler.go (hunk 6 above)
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go:79]: the node is used to determine the kernelVersion , and to determine kernel mapping (using kernel version). Then both kernel mapping(mld) and node are passed to the function that determines whether the module needs to run on the node or not

File: controllers/module_nmc_reconciler.go (hunk 7 above)
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go:96]: why to add another "if", if there is already code in the Join that takes care of it? This is the purpose of Join function: decide of there are errors in the array of errors , and create the appropriate error

File: controllers/module_nmc_reconciler.go (hunk 5 above)
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go]: yes, in this case it was a mistake

File: controllers/module_nmc_reconciler.go (hunk 7 above)
Comment (ybettan) [reply controllers/module_nmc_reconciler.go:96]: > why to add another "if", if there is already code in the Join that takes care of it?

Because it is not trivial for someone who didn't read the `Join` doc to understand those are ignored. Everywhere else in the code we are checking if an error is != `nil` before doing something with it except here.

If you think it is better this way, let's keep it, I just though it is a bit confusing.

File: internal/filter/filter.go [hunk 11]
```diff
@@ -169,6 +211,67 @@ func (f *Filter) FindModulesForNode(ctx context.Context, node client.Object) []r
...
+		reqs = append(reqs, reconcile.Request{NamespacedName: nsn})
+		reqSet.Insert(nsn)
+	}
+
+	nms, err := f.nmcHelper.Get(ctx, node.GetName())
+	if err != nil || nms == nil {
+		return reqs
+	}
+
+	// go over modules of NodeModulesConfig and add them to request if they are not there already
```
Comment (qbarrand) [inline internal/filter/filter.go:260]: Why should we do this? We have enqueued all Modules that we need to reconcile already.
Besides, this function should return NamespacedNames for Modules only, not Nodes or NMCs.

File: internal/filter/filter.go [hunk 12]
```diff
@@ -241,6 +344,16 @@ func (f *Filter) EnqueueAllPreflightValidations(ctx context.Context, mod client.
...
 
+func (f *Filter) FindModuleForJob(ctx context.Context, job client.Object) []reconcile.Request {
+	owner := metav1.GetControllerOfNoCopy(job)
+
+	if owner != nil && owner.Kind == "Module" {
+		req := reconcile.Request{NamespacedName: types.NamespacedName{Name: owner.Name, Namespace: job.GetNamespace()}}
+		return []reconcile.Request{req}
+	}
+	return nil
+}
```
Comment (qbarrand) [inline internal/filter/filter.go:355]: I do not think we need this function. Using `.Owns(&batchv1.Job{})` in `SetupWithManager` does this already.

File: controllers/module_nmc_reconciler.go (hunk 7 above)
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go:96]: i am keeping it: if we have a Golang library function in the code, i would expect that developer to read about it and understand

File: controllers/module_nmc_reconciler.go (hunk 10 above)
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go:139]: I prefer to set such function in a helper interface, makes it easier to write the unit-tests for the main reconciliation loop, and to check all the corner case in a dedicated unit-test for that specific function.

File: cmd/manager/main.go [hunk 13]
```diff
@@ -134,6 +136,7 @@ func main() {
 
 	daemonAPI := daemonset.NewCreator(client, constants.KernelLabel, scheme)
 	kernelAPI := module.NewKernelMapper(buildHelperAPI, sign.NewSignerHelper())
+	//nmcHandler := nmc.NewModuleNMCHandler(client, registryAPI, nmcHelper)
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply cmd/manager/main.go:139]: fixed

File: controllers/module_nmc_reconciler.go [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+
+	logger.Info("Starting Module-NMS reconcilation", "module name and namespace", req.NamespacedName)
+
+	mod, err := mnr.reconHelper.getRequestedModule(ctx, req.NamespacedName)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			logger.Info("Module deleted")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get the requested %s KMMO CR: %v", req.NamespacedName, err)
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go:67]: fixed

File: controllers/module_nmc_reconciler.go [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+		registryAPI: registryAPI,
+		nmcHelper:   nmcHelper,
+	}
+}
+
+func (mnrh *moduleNMCReconcilerHelper) getRequestedModule(ctx context.Context, namespacedName types.NamespacedName) (*kmmv1beta1.Module, error) {
+	mod := kmmv1beta1.Module{}
+
+	if err := mnrh.client.Get(ctx, namespacedName, &mod); err != nil {
+		return nil, fmt.Errorf("failed to get the kmmo module %s: %w", namespacedName, err)
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go:134]: fixed

File: controllers/module_nmc_reconciler.go [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+	}
+
+	opRes, err := controllerutil.CreateOrPatch(ctx, mnrh.client, nmc, func() error {
+		return mnrh.nmcHelper.SetModuleConfig(ctx, nmc, mld.Namespace, mld.Name, &moduleConfig)
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to enable module %s/%s in NMC %s: %v", mld.Namespace, mld.Name, nodeName, err)
+	}
+	logger.Info("enable module in NMC", "name", mld.Name, "namespace", mld.Namespace, "node", nodeName, "result", opRes)
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go:193]: fixed

File: controllers/module_nmc_reconciler.go [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+
+	opRes, err := controllerutil.CreateOrPatch(ctx, mnrh.client, nmc, func() error {
+		return mnrh.nmcHelper.RemoveModuleConfig(ctx, nmc, modNamespace, modName)
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to diable module %s/%s in NMC %s: %v", modNamespace, modName, nodeName, err)
+	}
+
+	logger.Info("disable module in NMC", "name", modName, "namespace", modNamespace, "node", nodeName, "result", opRes)
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go:216]: fixed

File: internal/filter/filter.go [hunk 18]
```diff
@@ -100,6 +127,21 @@ func (f *Filter) ModuleReconcilerNodePredicate(kernelLabel string) predicate.Pre
...
+		predicate.Or(nodeBecomesSchedulable, predicate.LabelChangedPredicate{}),
+	)
+}
+
+func ModuleNMCReconcileJobPredicate() predicate.Predicate {
+	return predicate.And(
+		skipDeletions,
+		skipCreations,
+		moduleJobSuccess,
+	)
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply internal/filter/filter.go:142]: Why do we need to process failed job? We are only interested in kicking in the reconciliation in case there is a chance that an image was built/created

File: internal/filter/filter.go [hunk 19]
```diff
@@ -169,6 +211,67 @@ func (f *Filter) FindModulesForNode(ctx context.Context, node client.Object) []r
...
+func (f *Filter) FindModulesForNMCNodeChange(ctx context.Context, node client.Object) []reconcile.Request {
+	logger := ctrl.LoggerFrom(ctx).WithValues("node", node.GetName())
+
+	reqs := make([]reconcile.Request, 0)
+
+	logger.Info("Listing all modules")
+
+	mods := kmmv1beta1.ModuleList{}
+
+	err := f.client.List(context.Background(), &mods)
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply internal/filter/filter.go:223]: fixed

File: internal/filter/filter.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (qbarrand) [reply internal/filter/filter.go:142]: Don't we want to log failed builds, at least?

File: internal/filter/filter.go [hunk 20]
```diff
@@ -75,6 +83,21 @@ var kmmClusterClaimChanged predicate.Predicate = predicate.Funcs{
...
 }
 
+var moduleJobSuccess predicate.Predicate = predicate.Funcs{
+	UpdateFunc: func(e event.UpdateEvent) bool {
+		job, ok := e.ObjectNew.(*batchv1.Job)
+		if !ok {
+			return true
+		}
+		owner := metav1.GetControllerOfNoCopy(job)
+		if owner != nil && owner.Kind == "Module" && job.Status.Succeeded == 1 {
```
[the remark this answers is not in the document]
Comment (qbarrand) [reply internal/filter/filter.go:93]: In fact I don't think we need to filter the owner kind, controller-runtime does that automatically if we use [`Owns`](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.15.1/pkg/builder#Builder.Owns).

File: internal/filter/filter.go [hunk 21]
```diff
@@ -169,6 +211,67 @@ func (f *Filter) FindModulesForNode(ctx context.Context, node client.Object) []r
 	return reqs
 }
 
+func (f *Filter) FindModulesForNMCNodeChange(ctx context.Context, node client.Object) []reconcile.Request {
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply internal/filter/filter.go:214]: fixed

File: internal/filter/filter.go [hunk 22]
```diff
@@ -169,6 +211,67 @@ func (f *Filter) FindModulesForNode(ctx context.Context, node client.Object) []r
...
+
+	for _, mod := range mods.Items {
+		logger := logger.WithValues("module name", mod.Name)
+
+		logger.V(1).Info("Processing module")
+
+		moduleSelectorMatchNode, err := utils.IsObjectSelectedByLabels(node.GetLabels(), mod.Spec.Selector)
+		if err != nil {
+			logger.Error(err, "could not determine if node is selected by module", "node", node.GetName(), "module", mod.Name)
+			return reqs
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply internal/filter/filter.go:241]: no, changing to continue

File: internal/filter/filter.go [hunk 23]
```diff
@@ -169,6 +211,67 @@ func (f *Filter) FindModulesForNode(ctx context.Context, node client.Object) []r
...
+		}
+
+		nsn := types.NamespacedName{Name: mod.Name, Namespace: mod.Namespace}
+
+		reqs = append(reqs, reconcile.Request{NamespacedName: nsn})
+		reqSet.Insert(nsn)
+	}
+
+	nms, err := f.nmcHelper.Get(ctx, node.GetName())
+	if err != nil || nms == nil {
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply internal/filter/filter.go:256]: Get function does not return an error in case NMC does not exists.Maybe i should change it to return NotFound instead of nil

File: internal/filter/filter.go (hunk 11 above)
Comment (yevgeny-shnaidman) [reply internal/filter/filter.go:260]: 1) because we should also find the modules that should be reconciled in the opposite direction ( they should not run on the node anymore. If Node's labels were changed/removed, the first part of the function will find the modules that should run on the node, but we must also check the already running Modules (stored in the NMC) , in case they need to be removed
2) the functions does return NamespacedName of the modules, which are stored in nmc.Spec.Modules

File: internal/filter/filter.go [hunk 24]
```diff
@@ -169,6 +211,71 @@ func (f *Filter) FindModulesForNode(ctx context.Context, node client.Object) []r
...
+		if reqSet.Has(nsn) {
+			continue
+		}
+		reqs = append(reqs, reconcile.Request{NamespacedName: nsn})
+	}
+
+	logger.Info("Adding reconciliation requests", "count", len(reqs))
+	logger.V(1).Info("New requests", "requests", reqs)
+
+	return reqs
```
Comment (qbarrand) [inline internal/filter/filter.go:276]: Could we use [`reqSet.UnsortedList()`](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets#Set.UnsortedList) instead? That way we don't need to maintain two separate collections.

File: internal/filter/filter.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (qbarrand) [reply internal/filter/filter.go:260]: Misread the `Get` implementation sorry. Let us leave it as is for now.

File: internal/filter/filter.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply internal/filter/filter.go:142]: That is not related to this controller, this one is responsible for module-nmc reconciliation. Controller responsible for building/signing will log what is needed

File: internal/filter/filter.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply internal/filter/filter.go:93]: i am keeping the filtering, but removing the mapping function

File: controllers/module_nmc_reconciler.go (hunk 3 above)
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go]: removed the "update"

File: internal/filter/filter.go (hunk 24 above)
Comment (yevgeny-shnaidman) [reply internal/filter/filter.go]: Fixed

File: internal/filter/filter.go [hunk 25]
```diff
@@ -100,6 +127,21 @@ func (f *Filter) ModuleReconcilerNodePredicate(kernelLabel string) predicate.Pre
 	)
 }
 
+func ModuleNMCReconcilerNodePredicate() predicate.Predicate {
+	return predicate.And(
+		skipDeletions,
```
Comment (qbarrand) [inline internal/filter/filter.go:132]: I think we'd want to delete a NMC if the corresponding node is gone.

File: internal/filter/filter.go [hunk 26]
```diff
@@ -169,6 +211,67 @@ func (f *Filter) FindModulesForNode(ctx context.Context, node client.Object) []r
...
 }
 
+// FindModulesForNMCNodeChange finds the modules that are affected by node changes that result
+// in ModuleNMCReconcilerNodePredicate predicate. First it find all the Module that can run on the node, based
+// on the Modules' Selector field and on node's labels. Then, in case NMC for the node exists, it adds all the
+// Modules already set in NMC ( in case they were not added in a previous step).
+func (f *Filter) FindModulesForNMCNodeChange(ctx context.Context, node client.Object) []reconcile.Request {
+	logger := ctrl.LoggerFrom(ctx).WithValues("node", node.GetName())
+
+	reqs := make([]reconcile.Request, 0)
```
Comment (qbarrand) [inline internal/filter/filter.go:221]: I think we can remove that variable since it's now unused.

File: internal/filter/filter.go (hunk 25 above)
Comment (yevgeny-shnaidman) [reply internal/filter/filter.go:132]: Currently i see it as an optimization (it does not affect the correctness of the code right now). So i would schedule it for the next PRs

File: internal/filter/filter.go (hunk 26 above)
Comment (yevgeny-shnaidman) [reply internal/filter/filter.go:221]: fixed

File: controllers/module_nmc_reconciler.go [hunk 27]
```diff
@@ -0,0 +1,241 @@
...
+				filter.ModuleNMCReconcilerNodePredicate(),
+			),
+		).
+		Watches(
+			&batchv1.Job{},
+			handler.EnqueueRequestForOwner(mgr.GetScheme(), mgr.GetRESTMapper(), &kmmv1beta1.Module{}, handler.OnlyControllerOwner()),
+			builder.WithPredicates(
+				filter.ModuleNMCReconcileJobPredicate(),
+			),
+		).
```
[the remark this answers is not in the document]
Comment (yevgeny-shnaidman) [reply controllers/module_nmc_reconciler.go:238]: No, Owns defines it own filter and handler. Since i would like to keep the filter, i am still using Watches
